scipy/spatial/ckdtree/src/ckdtree_cpp_methods.h

```diff
@@ -0,0 +1,193 @@
+
+#ifndef CKDTREE_CPP_METHODS
+#define CKDTREE_CPP_METHODS
+
+#ifdef CKDTREE_METHODS_IMPL
+#define CKDTREE_EXTERN extern "C"
+#else
+#define CKDTREE_EXTERN extern "C" 
+struct ckdtree;
+#endif 
+
+extern npy_float64 infinity;
+extern int number_of_processors;
+
+#ifndef NPY_LIKELY
+#define NPY_LIKELY(x) (x)
+#endif
+
+#include <cmath>
+
+#ifndef WIN32
+inline void 
+prefetch_datapoint(const npy_float64 *x, const npy_intp m) 
+{
+    const int cache_line = 64;  // x86, amd64
+    char *cur = (char*)x;
+    char *end = (char*)(x+m);
+    while (cur < end) { 
+        __builtin_prefetch((void*)cur);
+        cur += cache_line;
+    }
+}
+#else
+// On Windows, except when using MinGW
+#define prefetch_datapoint(x,y)
+#endif
+
+
+/*
+ * Utility functions
+ * =================
+ */
+ 
+inline npy_float64 
+dmax(const npy_float64 x, const npy_float64 y) 
+{
+    if (x > y)
+        return x;
+    else
+        return y;
+};
+
+inline npy_float64 
+dabs(const npy_float64 x)
+{
+    if (x > 0)
+        return x;
+    else
+        return -x;
+}
+
+
+/*
+ * Measuring distances
+ * ===================
+ */
+
+
+// XXX earlier versions of GCC and Clang might work
+#ifndef __INTEL_COMPILER
+#if ((defined(__GNUC__) && (__GNUC__ > 4))\
+    || (defined(__clang__) && ((__clang_major__ >= 3) && (__clang_minor >= 4)))\
+    || defined(__APPLE__)) // FIXME: clang defines don't work for clang 3.5 in Xcode
```

Comment:
I'm not sure... This came from a PR that was written for vectorizing spatial.distances. But thet PR was closed. I'll see if I can find it. It clearly works on clang 3.5 from Xcode on my Mac. I'll fix this. It should probably have a check on ```__GNUC_MINOR__``` as well. By the way, Intel C++ is able to vectorize the manually unrolled loop, but clang or gcc does not do it.